java-manta-client/src/main/java/com/joyent/manta/client/MantaClient.java

```diff
@@ -1493,46 +1499,80 @@ public void move(final String source, final String destination)
-                listObjects(source).forEach(mantaObject -> {
-                    try {
-                        String sourcePath = mantaObject.getPath();
-                        String relPath = sourcePath.substring(sourceDirPath.length());
-                        String destFullPath = destDirPath + SEPARATOR + relPath;
-
-                        if (mantaObject.isDirectory()) {
-                            move(mantaObject.getPath(), destFullPath);
-                        } else {
-                            putSnapLink(destFullPath, sourcePath, null);
-                        }
-                    } catch (IOException e) {
-                        throw new UncheckedIOException(e);
-                    }
-                });
-            } catch (UncheckedIOException e) {
-                throw e.getCause();
-            }
+        if (!existsAndIsAccessible(destinationDir)) {
+            putDirectory(destinationDir, true);
+        }
 
-            deleteRecursive(source);
-        } else {
-            putSnapLink(destination, source, null);
+        putSnapLink(destination, source, null);
+        delete(source);
+    }
+
+    /**
+     * Moves a directory from one path to another path. This operation is not
+     * transactional and will fail or produce inconsistent result if the source
+     * or the destination is modified while the operation is in progress.
+     *
+     * @param source Original path to move from
+     * @param destination Destination path to move to
+     * @param entry Directory supplemental data object
+     * @throws IOException thrown when something goes wrong
+     */
+    private void moveDirectory(final String source, final String destination,
+                               final MantaObjectResponse entry)
+            throws IOException {
+        putDirectory(destination, true);
+        MantaHttpHeaders sourceHeaders = entry.getHttpHeaders();
+        Long contentsCount = sourceHeaders.getResultSetSize();
+
+        // If we were just copying an empty directory, we just create the
+        // new directory and delete the original
+        if (contentsCount != null && contentsCount == 0L) {
             delete(source);
+            return;
         }
+
+        MantaObjectResponse destDir = head(destination);
+        String destDirPath = destDir.getPath();
+        String sourceDirPath = entry.getPath();
+
+        listObjects(source).forEach(mantaObject -> {
+            try {
+                String sourcePath = mantaObject.getPath();
+                String relPath = sourcePath.substring(sourceDirPath.length());
+                String destFullPath = destDirPath + SEPARATOR + relPath;
+
+                if (mantaObject.isDirectory()) {
+                    move(mantaObject.getPath(), destFullPath);
+                } else {
+                    putSnapLink(destFullPath, sourcePath, null);
+                }
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
```

Comment:
Doesn't this method already throw a checked IOException? Which I guess was the existing behavior :-/